_posts/2023-02-22-reveal-type-in-crystal.md

```diff
@@ -0,0 +1,397 @@
+
+Since it would not require a user defined macro/method it would not suffer from the issues exposed in the caveats.
+
+But maybe there is something intermediate we can do to allow more future use cases.
+
+In the `reveal_type` macro we needed to show the expression and its location. This is something that is already done in [`AST#raise`](https://crystal-lang.org/api/1.7.2/Crystal/Macros/ASTNode.html#raise%28message%29%3ANoReturn-instance-method). Unfortunately it’s not possible to tweak the output and it’s always treated as a compiler error:
+
+```crystal
+{% raw %}
+macro reveal_type(t)
+  {% t.raise "Lorem ipsum" %}
+end
+{% endraw %}
+```
+
+```console
+In program.cr:24:1
+
+  24 | reveal_type(config.smtp).host = "example.org"
+      ^----------
+Error: Lorem ipsum
+```
+
+If we would like to keep the `reveal_type` defined in user code I think it would be nice to have something similar to `AST#raise` to print information only. This could have the location, expression and `^-------` already solved, while allowing use to customize the message. Additionally,
+
+
+- It could allow multiple information invocation and not abort the compilation as `AST#raise` does.
+- It could have access to some additional information as final node type by having specific execution live-cycle: `AST#at_exit_info` for example.
+- It could be used to experiment with additional compile-time tooling (eg: checking if a database and model is up-to-date with one another)
+
+Some of these ideas resonate a lot with requests made by [paulcsmith](https://github.com/paulcsmith) of [Lucky](https://github.com/luckyframework/) on how to extend the compiler’s behavior. I expect that something like `AST#at_exit_info` or `AST#info` would be useful in that regard.
+
+## Conclusions
+
+The final shape of our solutions can be easily added in our Crystal app for development purposes.
+
+```crystal
+{% raw %}
+def reveal_type_helper(t : T, l) : T forall T
+  {% puts "   : #{T}" %}
+  t
+end
+
+macro reveal_type(t)
+  {% loc = "#{t.filename.id}:#{t.line_number}:#{t.column_number}" %}
+  {% puts "Revealed type #{loc.id}" %}
+  {% puts "  #{t.id}" %}
+  reveal_type_helper({{t}}, { {{loc.tr("/:.","___").id}}: 1 })
+end
+{% endraw %}
+```
+
+If we have an expression like `foo(bar.baz)` in which we are uncertain the type of `bar` we can:
+
+
+1. Surround `bar` with `reveal_type` as in `foo(reveal_type(bar).baz)`
+2. Build the program as usual.
+3. See the compiler output as
+
+```console
+Revealed type /path/to/program.cr:14:15
+  bar
+    : (String | Nil)
+```
+
+As mentioned before there are a couple of caveats with this solutions, but I think it works in the vast majority of cases. It was great to extend somehow the compiler tooling via user code only.
+
+It would be great to have some feedback if a proper built-it alternative of this would be valuable.
```

Comment:
```suggestion
It would be great to have some feedback if a proper built-in(not 100% sure if it should be built-it or built-in keeping room for suggestion)alternative to this would be valuable.
```